app-cpp/app355.cpp

```diff
@@ -393,6 +324,83 @@ void SatOrbit::OnSortOrbitalVectorMulti(orbit_iterator& inBegin, orbit_iterator&
+{
+    const std::size_t orbitVectorSize = ioOrbitalVector.size();
+
+        // calculate the number of TLEs per thread
+        const std::size_t orbitPerThread = orbitVectorSize / mNumThreads;
+        // Note: no need for mutex as threads do not contend for shared elements during sort
+        IteratorPairVector subListVector{};
+
+        // Block scope that has multiple threads quicksort segments of the vector
+        {
+            // create a list of threads the size of the number of threads specified
+            std::vector<std::thread> threadVector{};
+
+            for (std::size_t i = 0; i < mNumThreads; ++i)
+            {
+                // calculate the begin and end of the TLEs for the current thread
+                const auto orbitBegin = ioOrbitalVector.begin() + (i * orbitPerThread);
+                // if this is the last thread, add the remaining TLEs, else add orbitPerThread
+                const auto orbitEnd = (i >= mNumThreads - 1) ? ioOrbitalVector.end() : orbitBegin + orbitPerThread;
+                threadVector.push_back(std::thread(&SatOrbit::OnSortOrbitalVectorMulti, this, orbitBegin, orbitEnd));
+                
+                auto slice = std::make_tuple(orbitBegin, orbitEnd );
+                //IteratorPairVector::value_type slice{ orbitBegin, orbitEnd };
+                subListVector.push_back(std::move(slice));
+            }
+            
+            // Join the threads together
+            // wait for the threads to finish before returning
+            std::for_each(threadVector.begin(), threadVector.end(), [](std::thread& inThread)
+            {
+                inThread.join();
+            });
+        }
+
+        // Merge the sorted sublists
+        // Use multithreading by merging multilple sublists at a time
+        // Join the threads before merging the set of larger sublists
+        // Loop until there is only one list of completely sorted OrbitalData
+        while (subListVector.size() > 1)
+        {
+            // Merge sort the sublists 2 at a time
+            std::vector<std::thread> threadVector{};
+            IteratorPairVector newSubListVector{};
+
+            // This loop will merge the sublists 2 at a time, therefore we start at 1 and increment by 2
+            for (std::size_t i = 1; i < subListVector.size(); i += 2)
+            {
+                // Get first 2 sublists to merge
+                auto& [begin, mid] = subListVector[i-1];
+                auto& [mid2, end] = subListVector[i];
+                // Mids must be sequential in memory
+                assert(mid == mid2);
+
+                // Start a thread to merge the 2 sublists
+                threadVector.push_back(std::thread(&SatOrbit::OnSortMergeVector, this, begin, mid, end));
+                // Create a new sublist element from the 2 merged sublists onto a new list
+                newSubListVector.push_back(std::make_tuple(begin, end));
```

Comment:
👍 `std::make_tuple` as rvalue parameter.
fyi- but you could have also done:
```
    auto slice = std::make_tuple(begin, end); // note: non-const
    newSubListVector.push_back(std::move(slice)); // move instead of copy
```